redis/src/main/java/org/hibernate/ogm/datastore/redis/AbstractRedisDialect.java

```diff
@@ -316,4 +366,119 @@ protected void storeAssociation(
 			}
 		}
 	}
+
+	/**
+	 * Scan over keys. This method is aware whether the the client is connected to a Redis Cluster.
+	 * If so, then a Redis Cluster scan requires to iterate over master nodes and keep the
+	 * state within a custom {@link KeyScanCursor} instance.
+	 *
+	 * @param cursor
+	 * @param scanArgs
+	 *
+	 * @return
+	 */
+	protected KeyScanCursor<String> scan(KeyScanCursor<String> cursor, ScanArgs scanArgs) {
+
+		if ( !clusterMode ) {
+			return scan( connection, cursor, scanArgs );
+		}
+
+		return clusterScan( cursor, scanArgs );
+	}
+
+	private KeyScanCursor<String> scan(
+			RedisClusterCommands<String, String> commands,
+			KeyScanCursor<String> cursor,
+			ScanArgs scanArgs) {
+		if ( cursor != null ) {
+			return commands.scan( cursor, scanArgs );
+		}
+		return commands.scan( scanArgs );
+	}
+
+	@SuppressWarnings("unchecked")
+	private KeyScanCursor<String> clusterScan(KeyScanCursor<String> cursor, ScanArgs scanArgs) {
```

Comment:
At 25k Ops/sec slow gets relative. I didn't want to come up with async in the first place but it's possible. Lettuce is async under the hood so we could use `Future`s to query run queries on all masters. To control where a particular entity lives, we would need hash tagging but hash tagging is done by changing the key. Not sure whether I like the idea because it may lead to some surprise when looking directly into Redis.